packages/workflow-core/src/lib/workflow-runner.test.ts

```diff
@@ -1,34 +1,340 @@
+    const wf = createEventCollectingWorkflow({
+      definition: {
+        initial: 'initial',
+        states: {
+          initial: {
+            on: { EVENT: 'final' },
+          },
+          final: {
+            on: { EVENT: 'initial' },
+          },
+          alternative_initial: {
+            on: { EVENT: 'alternative_final' },
+          },
+          alternative_final: {
+            on: { EVENT: 'alternative_initial' },
+          },
+        },
+      },
+      workflowContext: { state: 'alternative_initial' },
+    });
+
+    wf.sendEvent({ type: 'EVENT', ...DEFAULT_PAYLOAD });
+
+    expect(wf.events).toStrictEqual([
+      { type: 'EVENT', state: 'alternative_final', ...DEFAULT_PAYLOAD },
+    ]);
   });
 
-  expect(service.state).toBe('inactive');
-  service.sendEvent({ type: 'TOGGLE' });
-  expect(service.state).toBe('active');
-  service.sendEvent({ type: 'TOGGLE' });
-  expect(service.state).toBe('inactive');
+  it('uses the last subscribed callback', () => {
+    const wf = new WorkflowRunner({
+      definition: TWO_STATES_MACHINE_DEFINITION,
+    });
+
+    const events = [];
+    wf.subscribe(event => events.push(event));
+    wf.subscribe(event => {});
+    wf.sendEvent({ type: 'EVENT', ...DEFAULT_PAYLOAD });
+
+    expect(events).toStrictEqual([]);
+  });
+
+  describe('transition plugins', () => {
+    describe('non blocking', () => {
+      it('does not allow to keep track of plugins running status using the callback', async () => {
+        const wf = createEventCollectingWorkflow({
+          definition: TWO_STATES_MACHINE_DEFINITION,
+          extensions: {
+            statePlugins: [
+              {
+                name: 'SuccessfulPlugin',
+                when: 'pre',
+                stateNames: ['initial'],
+                action: () => {},
+              },
+              {
+                name: 'FailingPlugin',
+                when: 'pre',
+                stateNames: ['initial'],
+                action: () => {
+                  throw new Error('some error');
+                },
+              },
+            ],
+          },
+        });
+
+        await wf.sendEvent({ type: 'EVENT', ...DEFAULT_PAYLOAD });
+        await sleep(1);
+
+        expect(wf.events).toStrictEqual([
+          { type: 'EVENT', payload: { some: 'payload' }, state: 'final' },
+        ]);
+      });
+
+      it('does not fail transitions', async () => {
+        const wf = createEventCollectingWorkflow({
+          definition: TWO_STATES_MACHINE_DEFINITION,
+          extensions: {
+            statePlugins: [
+              {
+                name: 'FailingPrePlugin',
+                when: 'pre',
+                stateNames: ['initial'],
+                action: () => {
+                  throw new Error();
+                },
+              },
+            ],
+          },
+        });
+
+        await wf.sendEvent({ type: 'EVENT', ...DEFAULT_PAYLOAD });
+        await sleep(1);
```

Comment:
Why do we need the `sleep`s?

Reply:
Now I can see that `__handleAction` is async and not being awaited, might be the source of then issue